File: src/main/java/com/somefrills/misc/SortedList.java
package com.somefrills.misc;

import org.jspecify.annotations.NonNull;

import java.util.*;

public class SortedList<T> implements List<T> {
    private T[] array;

    public SortedList(T[] array) {
        this.array = Arrays.copyOf(array, array.length);
        Arrays.sort(this.array);
    }

    @SuppressWarnings("unchecked")
    public SortedList(List<T> list) {
        this.array = (T[]) list.toArray();
        Arrays.sort(this.array);
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    public SortedList() {
        this.array = (T[]) new Comparable[0];
    }

    @Override
    public int size() {
        return array.length;
    }

    @Override
    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public boolean contains(Object o) {
        if (o == null) return false;
        try {
            return Arrays.binarySearch(array, o) >= 0;
        } catch (ClassCastException e) {
            return false;
        }
    }

    @Override
    public @NonNull Iterator<T> iterator() {
        return Arrays.asList(array).iterator();
    }

    @Override
    public @NonNull Object[] toArray() {
        return Arrays.copyOf(array, array.length);
    }

    @Override
    @SuppressWarnings("unchecked")
    public @NonNull <T1> T1[] toArray(@NonNull T1[] a) {
        if (a.length < array.length) {
            return (T1[]) Arrays.copyOf(array, array.length, a.getClass());
        }
        System.arraycopy(array, 0, a, 0, array.length);
        if (a.length > array.length) a[array.length] = null;
        return a;
    }

    @Override
    public boolean add(T t) {
        int index = Arrays.binarySearch(array, t);
        if (index < 0) index = -index - 1;

        array = insertAt(array, t, index);
        return true;
    }

    @Override
    public boolean remove(Object o) {
        int index = indexOf(o);
        if (index < 0) return false;
        array = removeAt(array, index);
        return true;
    }

    @Override
    public boolean containsAll(@NonNull Collection<?> c) {
        for (Object o : c) {
            if (!contains(o)) return false;
        }
        return true;
    }

    @Override
    public boolean addAll(@NonNull Collection<? extends T> c) {
        boolean modified = false;
        for (T t : c) {
            add(t);
            modified = true;
        }
        return modified;
    }

    @Override
    public boolean addAll(int index, @NonNull Collection<? extends T> c) {
        // index is ignored because we maintain sorted order
        return addAll(c);
    }

    @Override
    public boolean removeAll(@NonNull Collection<?> c) {
        boolean modified = false;
        for (Object o : c) {
            modified |= remove(o);
        }
        return modified;
    }

    @Override
    public boolean retainAll(@NonNull Collection<?> c) {
        List<T> temp = new ArrayList<>();
        for (T t : array) {
            if (c.contains(t)) temp.add(t);
        }
        if (temp.size() != array.length) {
            array = temp.toArray(array.clone());
            return true;
        }
        return false;
    }

    @Override
    public void clear() {
        array = Arrays.copyOf(array, 0);
    }

    @Override
    public T get(int index) {
        return array[index];
    }

    @Override
    public T set(int index, T element) {
        T old = array[index];
        remove(index);
        add(element);
        return old;
    }

    @Override
    public void add(int index, T element) {
        // ignore index to preserve sorting
        add(element);
    }

    @Override
    public T remove(int index) {
        T old = array[index];
        array = removeAt(array, index);
        return old;
    }

    @Override
    public int indexOf(Object o) {
        if (o == null) return -1;
        try {
            int idx = Arrays.binarySearch(array, o);
            return idx >= 0 ? idx : -1;
        } catch (ClassCastException e) {
            return -1;
        }
    }

    @Override
    public int lastIndexOf(Object o) {
        return indexOf(o); // no duplicates handling improvement
    }

    @Override
    public @NonNull ListIterator<T> listIterator() {
        return Arrays.asList(array).listIterator();
    }

    @Override
    public @NonNull ListIterator<T> listIterator(int index) {
        return Arrays.asList(array).listIterator(index);
    }

    @Override
    public @NonNull List<T> subList(int fromIndex, int toIndex) {
        return Arrays.asList(Arrays.copyOfRange(array, fromIndex, toIndex));
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof List<?> other)) return false;
        return Arrays.equals(this.array, other.toArray());
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(array);
    }

    @Override
    public String toString() {
        return stream().map(s -> Utils.capitalizeType(s.toString())).reduce((a, b) -> a + ", " + b).orElse("");
    }

    private T[] insertAt(T[] arr, T value, int index) {
        arr = Arrays.copyOf(arr, arr.length + 1);
        System.arraycopy(arr, index, arr, index + 1, arr.length - index - 1);
        arr[index] = value;
        return arr;
    }

    private T[] removeAt(T[] arr, int index) {
        T[] newArr = Arrays.copyOf(arr, arr.length - 1);
        System.arraycopy(arr, 0, newArr, 0, index);
        System.arraycopy(arr, index + 1, newArr, index, arr.length - index - 1);
        return newArr;
    }
}

File: src/main/kotlin/com/somefrills/misc/SortedList.kt
package com.somefrills.misc

import com.somefrills.utils.TextUtils
import java.util.*

class SortedList<T : Comparable<T>>(private var array: Array<T>) : MutableList<T> {

    @Suppress("UNCHECKED_CAST")
    constructor(list: List<T>) : this(
        (arrayOfNulls<Comparable<T>>(list.size) as Array<T>).also { arr ->
            list.forEachIndexed { i, item -> arr[i] = item }
        }
    ) {
        array.sort()
    }

    constructor(list: SortedList<T>) : this(list.array.copyOf())

    @Suppress("UNCHECKED_CAST")
    constructor() : this(arrayOfNulls<Comparable<T>>(0) as Array<T>)

    init {
        array.sort()
    }

    override val size: Int
        get() = array.size

    override fun isEmpty(): Boolean = size == 0

    override fun contains(element: T): Boolean {
        return try {
            array.binarySearch(element) >= 0
        } catch (_: ClassCastException) {
            false
        }
    }

    override fun containsAll(elements: Collection<T>): Boolean {
        return elements.all { contains(it) }
    }

    override fun get(index: Int): T = array[index]

    override fun indexOf(element: T): Int {
        return try {
            val idx = array.binarySearch(element)
            if (idx >= 0) idx else -1
        } catch (_: ClassCastException) {
            -1
        }
    }

    override fun lastIndexOf(element: T): Int = indexOf(element)

    override fun iterator(): MutableIterator<T> = array.toMutableList().iterator()

    override fun listIterator(): MutableListIterator<T> = array.toMutableList().listIterator()

    override fun listIterator(index: Int): MutableListIterator<T> = array.toMutableList().listIterator(index)

    override fun subList(fromIndex: Int, toIndex: Int): MutableList<T> {
        return array.slice(fromIndex until toIndex).toMutableList()
    }

    override fun add(element: T): Boolean {
        var index = array.binarySearch(element)
        if (index < 0) index = -index - 1
        array = insertAt(array, element, index)
        return true
    }

    override fun add(index: Int, element: T) {
        add(element)
    }

    override fun addAll(elements: Collection<T>): Boolean {
        var modified = false
        for (element in elements) {
            add(element)
            modified = true
        }
        return modified
    }

    override fun addAll(index: Int, elements: Collection<T>): Boolean {
        return addAll(elements)
    }

    override fun clear() {
        @Suppress("UNCHECKED_CAST")
        array = arrayOfNulls<Comparable<T>>(0) as Array<T>
    }

    override fun remove(element: T): Boolean {
        val index = indexOf(element)
        return if (index < 0) false else {
            array = removeAt(array, index)
            true
        }
    }

    override fun removeAll(elements: Collection<T>): Boolean {
        var modified = false
        for (element in elements) {
            modified = modified or remove(element)
        }
        return modified
    }

    override fun removeAt(index: Int): T {
        val old = array[index]
        array = removeAt(array, index)
        return old
    }

    override fun retainAll(elements: Collection<T>): Boolean {
        val temp = array.filter { it in elements }
        @Suppress("UNCHECKED_CAST")
        val tempArray = arrayOfNulls<Comparable<T>>(temp.size).also { arr ->
            temp.forEachIndexed({ i, item -> arr[i] = item as Comparable<T> })
        } as Array<T>
        return if (temp.size != array.size) {
            array = tempArray
            true
        } else {
            false
        }
    }

    override fun set(index: Int, element: T): T {
        val old = array[index]
        removeAt(index)
        add(element)
        return old
    }

    override fun equals(other: Any?): Boolean {
        if (other !is List<*>) return false
        return array.contentEquals(other.toTypedArray())
    }

    override fun hashCode(): Int = array.contentHashCode()

    override fun toString(): String {
        return array.asSequence()
            .map { TextUtils.capitalizeType(it.toString()) }
            .reduceOrNull { a, b -> "$a, $b" }
            ?: ""
    }

    private fun insertAt(arr: Array<T>, value: T, index: Int): Array<T> {
        @Suppress("UNCHECKED_CAST")
        val newArr = arr.copyOf(arr.size + 1) as Array<T>
        System.arraycopy(arr, index, newArr, index + 1, arr.size - index)
        newArr[index] = value
        return newArr
    }

    private fun removeAt(arr: Array<T>, index: Int): Array<T> {
        @Suppress("UNCHECKED_CAST")
        val newArr = arr.copyOf(arr.size - 1) as Array<T>
        System.arraycopy(arr, 0, newArr, 0, index)
        System.arraycopy(arr, index + 1, newArr, index, arr.size - index - 1)
        return newArr
    }
}

